CryptoTechReminderSystem/Gateway/HarvestGateway.cs

```diff
@@ -102,17 +110,39 @@ private JObject RetrieveFromEndPoint(string address)
             return response.Result;
         }
 
-        private JObject RetrieveWithPagination(string address)
+        private async Task<JObject> GetApiResponse(string address)
+        {   
+            int THROTTLE_TIME_IN_MS = 150;
+            int MAX_ATTEMPTS = 5;
+            
+            var harvestRetryPolicy = Policy
+                .HandleResult<HttpResponseMessage>(r => r.StatusCode == System.Net.HttpStatusCode.TooManyRequests)
+                .WaitAndRetryAsync(
+                    MAX_ATTEMPTS, attempt => TimeSpan.FromMilliseconds(THROTTLE_TIME_IN_MS * Math.Pow(2, attempt)));
+            
+            var response = await harvestRetryPolicy.ExecuteAsync(() => (_client.GetAsync(address)));
+                
+            return JObject.Parse(await response.Content.ReadAsStringAsync());
```

Comment:
I'm curious about how you decided on the throttle time and max attempts 😄 

I'm also curious about the `Math.Pow` bit on line 121. Could you explain why you're using that in the `TimeSpan.FromMilliseconds` method? 😄

Reply:
Thanks for taking a look Dushan 😀

Re: throttle time, I chose to start at the Harvest [rate limit](https://help.getharvest.com/api-v2/introduction/overview/general/#rate-limiting) interval (100 requests per 15s == 1 request per 150ms).

The `Math.pow` bit is the exponential backoff, so it should start by retrying 150ms from the first request and increase the gap between requests exponentially.  But re-reading this I've noticed it's more than exponential because it multiplies the output of the exponentiation by THROTTLE_TIME_IN_MS 🤦🏻 .  I'll change that to:

```
MAX_ATTEMPTS, attempt => TimeSpan.FromMilliseconds(Math.Pow(2, attempt)));
```

Five max attempts is arbitrary...  I chose 5 because The Wolves' use cases aren't time-sensitive within, say, 5 - 10 minutes.  But after your question I calculated what the request interval would be after 5 attempts with exponential backoff starting from 150ms and it's... longer than I've been alive 😂

Thanks for sending me on this journey Dushan 😂   I don't know what best practice is here.  Do you think we should reduce the initial throttle time as well as the max retry attempts?  WolframAlpha [shows](https://www.wolframalpha.com/input/?i=%28%28150%29%5E2%29+milliseconds+in+minutes) a 150ms initial throttle with 2 max retry attempts results in a max interval of 22.5 seconds.  What do you reckon?

Reply:
What's wrong with just retrying X number of times or once per minute or both? As you say, the use case isn't that time critical. What's most likely to happen is you have a serious issue and your max retries get hit, or something is resolved and allows you to succeed but the exponential backoff is now long enough it needs to be restarted. I'd personally just keep it simple.

Reply:
I'm in agreement with Steve here, I'd be inclined to keep it simple and we can iterate on it if the need arises in the future. As for what to choose for the simple option I'm happy leave that choice to you - though I'd make sure you add the reason for your choice to the commit message in case anyone looks into the history in the future 😄

Reply:
I agree too, thanks for probing this one.  I've tagged you in the latest commit so will resolve this convo.